generator/src/codegen/cpp.rs

```diff
@@ -0,0 +1,385 @@
+use super::Codegen;
+use super::flat_ast::*;
+use std::io::Result;
+use ::heck::*;
+
+pub struct Generator<'a> {
+    output_dir: &'a std::path::Path,
+    select: Select,
+    hwriter: Option<::writer::Writer<std::fs::File>>,
+    cwriter: Option<::writer::Writer<std::fs::File>>,
+    version: &'a str,
+    stem: &'a str
+}
+
+enum Select {
+    H,
+    CPP
+}
+
+impl<'a> Generator<'a> {
+    pub fn new(output_dir: &'a std::path::Path, stem: &'a str, version: &'a str) -> Self {
+        Self {
+            output_dir,
+            select: Select::H,
+            hwriter: None,
+            cwriter: None,
+            version,
+            stem
+        }
+    }
+
+    fn convert_basic_types(type_: &str) -> Option<&'static str> {
+        match type_ {
+            "char" => Some("char"),
+            "u8" => Some("uint8_t"),
+            "i8" => Some("int8_t"),
+            "u16" => Some("uint16_t"),
+            "i16" => Some("int16_t"),
+            "u32" => Some("uint32_t"),
+            "i32" => Some("int32_t"),
+            "u64" => Some("uint64_t"),
+            "i64" => Some("int64_t"),
+            "float" => Some("float"),
+            "double" => Some("double"),
+            "string" => Some("std::string"),
+            "bool" => Some("bool"),
+
+            _ => None
+        }
+    }
+
+    fn select(&mut self, s: Select) {
+        self.select = s;
+    }
+
+    fn indent(&mut self) {
+        match self.select {
+            Select::H => self.hwriter.as_mut().unwrap().indent(),
+            Select::CPP => self.cwriter.as_mut().unwrap().indent()
+        };
+    }
+
+    fn dedent(&mut self) {
+        match self.select {
+            Select::H => self.hwriter.as_mut().unwrap().dedent(),
+            Select::CPP => self.cwriter.as_mut().unwrap().dedent()
+        };
+    }
+
+    fn write(&mut self, val: impl AsRef<str>) -> Result<&mut Self> {
+        match self.select {
+            Select::H => self.hwriter.as_mut().unwrap().write(val)?,
+            Select::CPP => self.cwriter.as_mut().unwrap().write(val)?
+        };
+        Ok(self)
+    }
+
+    fn generate_header(&mut self) -> Result<String> {
+        let filename = self.output_dir.to_str().unwrap().to_owned() + &format!("{}.h", self.stem);
```

Comment:
Could probably simplify this: `let filename = format!("{}{}.h", self.output_dir.to_str().unwrap(), self.stem);`.